WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -9,46 +9,176 @@ import Alamofire
+    private let authenticator: ((URL) throws(AuthenticationError) -> URL)?
+
+    init(
+        coreDataStack: CoreDataStackSwift = ContextManager.shared,
+        authenticator: ((URL) throws(AuthenticationError) -> URL)? = nil
+    ) {
+        self.coreDataStack = coreDataStack
+        self.authenticator = authenticator
     }
 
+    /// Sign in WP.com account.
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
     @MainActor
-    func signIn(from viewController: UINavigationController) async {
+    func signIn(from viewController: UIViewController, context: SignInContext) async -> TaggedManagedObjectID<WPAccount>? {
+        WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "start"])
+        do {
+            let account = try await attemptSignIn(from: viewController, context: context)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "success"])
+            return account
+        } catch {
+            present(error, from: viewController)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "error", "error": "\(error)"])
+            return nil
+        }
+    }
+
+    /// Sign in WP.com account.
+    ///
+    /// - SeeAlso: `signIn`
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
+    @MainActor
+    func attemptSignIn(from viewController: UIViewController, context: SignInContext) async throws(SignInError) -> TaggedManagedObjectID<WPAccount> {
+        let defaultAccount = try? WPAccount.lookupDefaultWordPressComAccount(in: coreDataStack.mainContext)
+        let hasAlreadySignedIn = defaultAccount != nil
+
         let token: String
         do {
-            token = try await authenticate(from: viewController)
+            token = try await authenticate(from: viewController, prefersEphemeralWebBrowserSession: hasAlreadySignedIn)
         } catch {
-            if let error = error as? WordPressDotComAuthenticator.Error {
-                presentSignInError(error, from: viewController)
-            } else {
-                wpAssertionFailure("WP.com web login failed", userInfo: ["error": "\(error)"])
-            }
-            return
+            throw .authentication(error)
         }
 
-        let delegate = WordPressAuthenticator.shared.delegate!
-        let credentials = AuthenticatorCredentials(wpcom: WordPressComCredentials(authToken: token, isJetpackLogin: false, multifactor: false))
         SVProgressHUD.show()
-        delegate.sync(credentials: credentials) {
+        defer {
             SVProgressHUD.dismiss()
+        }
+
+        // Fetch WP.com account details
+        let user: RemoteUser
+        do {
+            let service = AccountServiceRemoteREST(wordPressComRestApi: .defaultApi(oAuthToken: token, userAgent: WPUserAgent.defaultUserAgent()))
```

Comment:
Is there a need to customize the user agent as opposed to using the  default parameter (`WPUserAgent.wordPress`)?

Reply:
(nit) Extract this to a separate method to hide the details (`withCheckedThrowingContinuation`, mapping an error).